### PR TITLE
DRPC

### PR DESCRIPTION
**Pull request by akshaisarma**
This builds off the initial DRPC work from @ShriramKumar and redoes it without replicating the Storm code. Feel free to take a look but I'll wire this up with an actual instance to test first.

### CODE REVIEW

File: src/main/java/com/yahoo/bullet/storm/BulletStormConfig.java [hunk 1]
```diff
@@ -62,16 +67,21 @@
...
+    /**
+     * Constructor that loads the defaults and augments it with defaults.
+     *
+     * @param other The other config to wrap.
+     */
+    public BulletStormConfig(Config other) {
+        // Load Bullet and Storm defaults. Then merge the other.
+        super(DEFAULT_STORM_CONFIGURATION);
+        merge(other);
+        log.info("Merged settings:\n {}", getAll(Optional.empty()));
```
Comment (hbxie) [inline src/main/java/com/yahoo/bullet/storm/BulletStormConfig.java]: I saw this change in your previous PRs. Will you delete those from your previous PRs?

File: src/main/java/com/yahoo/bullet/storm/drpc/DRPCQueryResultPubscriber.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+     * and {@link DRPCConfig#DRPC_HTTP_PATH} to be set. To be used in PubSub.Context#QUERY_SUBMISSION mode.
+     *
+     * @param config A non-null config that contains the required settings.
+     */
+    public DRPCQueryResultPubscriber(BulletConfig config) {
+        Objects.requireNonNull(config);
+
+        Number connectTimeout = config.getRequiredConfigAs(DRPCConfig.DRPC_HTTP_CONNECT_TIMEOUT_MS, Number.class);
+        Number retryLimit = config.getRequiredConfigAs(DRPCConfig.DRPC_HTTP_CONNECT_RETRY_LIMIT, Number.class);
+        List<String> urls = (List<String>) config.getRequiredConfigAs(DRPCConfig.DRPC_SERVERS, List.class);
```
Comment (hbxie) [inline src/main/java/com/yahoo/bullet/storm/drpc/DRPCQueryResultPubscriber.java]: This local var name is the same as line 38 and make it a little bit hard to read. Can you change to a different name?

File: src/main/resources/bullet_storm_defaults.yaml [hunk 3]
```diff
@@ -61,19 +61,19 @@ bullet.topology.metrics.classes:
...
 bullet.topology.filter.bolt.cpu.load: 100.0
 bullet.topology.filter.bolt.memory.on.heap.load: 256.0
 bullet.topology.filter.bolt.memory.off.heap.load: 160.0
-bullet.topology.filter.bolt.parallelism: 35
+bullet.topology.filter.bolt.parallelism: 32
 bullet.topology.join.bolt.cpu.load: 100.0
 bullet.topology.join.bolt.memory.on.heap.load: 512.0
 bullet.topology.join.bolt.memory.off.heap.load: 160.0
-bullet.topology.join.bolt.parallelism: 20
+bullet.topology.join.bolt.parallelism: 2
```
Comment (hbxie) [inline src/main/resources/bullet_storm_defaults.yaml]: The numbers are changed a lot. Is there any reason for it?

File: src/main/java/com/yahoo/bullet/storm/drpc/DRPCQueryResultPubscriber.java (hunk 2 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/storm/drpc/DRPCQueryResultPubscriber.java]: Will do

File: src/main/resources/bullet_storm_defaults.yaml (hunk 3 above)
Comment (akshaisarma) [reply src/main/resources/bullet_storm_defaults.yaml]: For defaults, these were too high relative to the filter bolt parallelisms, so I dropped it. This parallelism is sufficient for hundreds of simultaneous queries.

File: src/main/java/com/yahoo/bullet/storm/QuerySpout.java [hunk 4]
```diff
@@ -23,36 +30,51 @@
...
+        }
     }
 
     @Override
     public void nextTuple() {
         PubSubMessage message = null;
         try {
             message = subscriber.receive();
-        } catch (PubSubException e) {
+        } catch (Exception e) {
```
Comment (hbxie) [inline src/main/java/com/yahoo/bullet/storm/QuerySpout.java]: What else exception does it throw? I checked the interface definition, only PubSubException thrown.

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/storm/QuerySpout.java]: To catch RuntimeExceptions. Basically if anything bad happens for any reason, we don't want to restart the spout (ex. bad JSON in PubSubMessage throws a GSON RuntimeException)

File: src/main/java/com/yahoo/bullet/storm/drpc/DRPCQuerySubscriber.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        PubSubMessage message = new PubSubMessage(id, content, new Metadata(null, returnInfo), sequence);
+
+        emittedIDs.put(ImmutablePair.of(id, sequence), drpcID);
+        return Collections.singletonList(message);
+    }
+
+    @Override
+    public void close() {
+        log.warn("Failing all pending requests: {}", emittedIDs);
+        emittedIDs.values().forEach(spout::fail);
```
Comment (NathanSpeidel) [inline src/main/java/com/yahoo/bullet/storm/drpc/DRPCQuerySubscriber.java]: As we discussed, messages are not removed from emittedIDs when a message is ack'ed...  Override commit() and do that?

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/storm/drpc/DRPCQuerySubscriber.java]: Done

**Comment by akshaisarma** [comment]
I tested this out on a Storm cluster with a custom build of bullet-service using the DRPC pubsub. Looks to be working.

**Pull request merged**